### PR TITLE
Fix for parameter "%n" on shared folders

### PR DESCRIPTION
**Pull request by k1l1**
Fix to allow '%n' parameter to give out real relative file path in shared folder scenario. 
Addressing issues: #48 and possibly #38 

### Scenario
User **A** creates Folder _/myfolder_ at shares it with User **B**
User **B** moves Folder from root dir to _/subfolder/myfolder_
User **B** creates file _/subfolder/myfolder/file1_ in this folder which triggers a workflow script:

**Result before:**
user-id (%o): A
nextcloud-relative path (%n): _B/files/subfolder/myfolder/file1_ **<---- file does not exists on drive at this position**

**Result after:**
user-id (%o): A
nextcloud-relative path (%n): _B/files/myfolder/file1_ **<---- file exists**

### CODE REVIEW

**Comment by Pingger** [comment]
> Addressing issues: #48 and possibly #38

Issue #38 is not adressed.

File: lib/Operation.php [hunk 1]
```diff
@@ -73,7 +75,18 @@ protected function buildCommand(string $template, Node $node, string $event, arr
...
 		if (strpos($command, '%n')) {
 			// Nextcloud relative-path
-			$command = str_replace('%n', escapeshellarg($node->getPath()), $command);
+                        $nodeID = -1;
+                        try {
+                                $nodeID = $node->getId();
+                        } catch (InvalidPathException $e) {
+                        } catch (NotFoundException $e) {
+                        }
+                        $config = new Config('config/');
```
Comment (blizzz) [inline lib/Operation.php:84]: this should come from \OCP\IConfig (method `getSystemValue`) and ought to be injected in the constructor.

File: lib/Operation.php [hunk 2]
```diff
@@ -73,7 +75,18 @@ protected function buildCommand(string $template, Node $node, string $event, arr
...
+                        $base_path = $config->getValue('datadirectory');
+
+                        $path = Filesystem::getLocalFile(Filesystem::getPath($nodeID));
```
Comment (blizzz) [inline lib/Operation.php:87]: $nodeID is not necessarily set, since the exceptions are suppressed and not handled further

**Review by blizzz — CHANGES_REQUESTED** [review]

**Comment by wiswedel** [comment]
@k1l1 Can you work with @blizzz 's feedback or do you need help resolving the open issues?

**Comment by k1l1** [comment]
I will try it myself this week and we'll see how it goes. I have to note that I have basically not much experience in php and the nextcloud environment, so please be gentle with me ;)

**Comment by wiswedel** [comment]
> so please be gentle with me

Sure thing. Thanks for your contribution in the first place. Don't hesitate to ask if you're reaching dead ends. We're here to help.

File: lib/Operation.php (hunk 1 above)
Comment (k1l1) [reply lib/Operation.php]: Should be resolved. Can you check it?

File: lib/Operation.php (hunk 2 above)
Comment (k1l1) [reply lib/Operation.php]: Hmm, what are the cases where $nodeID is not not set and what is the best way to handle the exception? Should we fallback to the old solution (and have a possibly wrong file path on shared folders), or leave the %n path empty?

**Comment by k1l1** [comment]
I can also confirm that creating and renaming files in group folders works and triggers a **workflow_script**
Is this the behavior which is missing for you?  @Pingger


**Steps**
1. create _user1_ and _user2_ and set them to group _standard_
2. Create groupfolder _gf_test_ set it to group _standard_
3. Create flow on create file When **File created** and **FILE MIME type**  **is** **PDF documents**
4. Upload pdf to _gf_test/sample.pdf_ with _user1_, workflow script gives for **%n** now __groupfolders/1/sample.pdf_
4. Upload pdf to _gf_test/sample2.pdf_ with _user2_, workflow script gives for **%n** now __groupfolders/1/sample2.pdf_

The path __groupfolders/1/sample.pdf_ exists in nextclouds data folder!

File: lib/Operation.php (hunk 2 above)
Comment (wiswedel) [reply lib/Operation.php]: @blizzz

**Review by blizzz — APPROVED** [review]
Looks good and does what is advertized :)

I added another commit fixing whitespaces – [we use tabs not spaces](https://docs.nextcloud.com/server/20/developer_manual/getting_started/codingguidelines.html)

**Comment by blizzz** [comment]
@k1l1 one last thing: would you sign off your commits? https://github.com/nextcloud/workflow_script/pull/64/checks?check_run_id=1447292190

**Comment by blizzz** [comment]
and apologies for my late reaction, the notification drowned among others :(

**Comment by k1l1** [comment]
merged!

**Comment by jakobroehrl** [comment]
@blizzz When do you plan to relase this fix? 
Would love to test it :)

**Comment by wiswedel** [comment]
> @blizzz When do you plan to relase this fix?
> Would love to test it :)

@jakobroehrl release is out, thanks for your patience: https://github.com/nextcloud/workflow_script/releases/tag/v1.3.3

**Comment by jakobroehrl** [comment]
Thanks for the information. 
Tested it already and it works great! :)

**Pull request merged**